(DOCSP-26313): Swift: Add a Handle Sync Errors page for SwiftUI (#2524)
mongodb/docs-realm: source/sdk/swift/swiftui/handle-sync-errors.txt

For a SwiftUI-friendly implementation of a Sync error handler, create 
an ``ObservableObject`` with an optional ``@Published`` variable to contain
a potential error. This handler uses the :swift-sdk:`SyncManager 
<Typealiases.html#/s:10RealmSwift11SyncManagera>` to listen for errors.
All errors reported by an error handler registered on the ``SyncManager`` 
are of the type ``SyncError``.

For more information, refer to the underlying Objective-C :objc-sdk:`RLMSyncError 
<Enums/RLMSyncError.html>`.

.. literalinclude:: /examples/generated/swiftui/HandleSyncErrors.snippet.swiftui-error-handler.swift
   :language: swift

.. include:: /includes/sync-errors-in-app-services.rst

Initialize the error handler as a ``@StateObject``. Inject it into the 
view hierarchy as an environment object. In this example, we display an 
``.alert`` to the user when a Sync error occurs.

.. literalinclude:: /examples/generated/swiftui/HandleSyncErrors.snippet.swiftui-app-with-error-handler.swift
   :language: swift

Then, in the view where you are observing the Realm ``App``, you can use the 
error handler as an ``@EnvironmentObject`` to react to Sync errors. An 

> error that occurs here pops up an alert for the user, thanks to the ``.alert``

Reviewer: ```suggestion
error that occurs here pops up an alert for the user due to the ``.alert``
```

> set in the view above. To see this in action, if you're following along
> with the template app, try registering a user that already exists.

Reviewer: ```suggestion
set in the view above.
```

i would cut the stuff abt template app b/c you didn't tell reader earlier to be following template app. it was only mentioned in an admonition.

plus, don't think it's nec to understand this code example.

   :language: swift

.. _swiftui-handle-client-reset-errors:

Handle Client Reset Errors
--------------------------

A client reset error is a specific type of Sync error that occurs when the 
realm on the user's device can no longer sync with the backend. For more
information about the underlying error and recovery options, refer to:
:ref:`ios-client-reset`.

Realm's SwiftUI property wrappers ``@AsyncOpen`` and ``@AutoOpen`` enable
you to handle a client reset error by adding a few things to your app:

- Add client reset handling information to the :swift-sdk:`flexibleSyncConfiguration() 
  <Extensions/User.html#/s:So7RLMUserC10RealmSwiftE25flexibleSyncConfigurationAC0B0V0F0VyF>`
- Force an ``@AsyncOpen`` or ``@AutoOpen`` view to reload after a client reset is complete

Add Client Reset Handling Information to the Flexible Sync Configuration
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The Swift SDK can handle a client reset with different behaviors depending
on the needs of your app. For a full list of client reset modes and behaviors,
refer to :ref:`ios-client-resets`.

> To automatically handle a client reset, set the ``clientResetMode`` parameter 

Reviewer: i think this is an argument, not a parameter (unless there's smthn swift specific here). alternatively, you could say 'property'

Author: Hmm. The internets say "A parameter is the variable listed inside the parentheses in the function definition. An argument is the value that is sent to the function when it is called." I interpret this to mean that `clientResetMode` is a parameter, because it's a variable in the [SyncConfiguration definition](https://www.mongodb.com/docs/realm-sdks/swift/latest/Structs/SyncConfiguration.html#/), while `.recoverUnsyncedChanges` used in this example would be an argument. (The API docs also refer to them as parameters so I assumed that was the correct terminology to use in this case.)

Reviewer: ok, will defer to API docs.

in your :swift-sdk:`flexibleSyncConfiguration() 
<Extensions/User.html#/flexibleSyncConfiguration(clientResetMode:cancelAsyncOpenOnNonFatalErrors:)>`.
You can review the possible values and their behaviors in the 
:swift-sdk:`ClientResetMode enum <Enums/ClientResetMode.html>`

> In this example, we set the ``clientResetMode`` to ``.recoverUnsyncedChanges``.

Reviewer: why not `recoverOrDiscardUnsyncedChanges` since that's what's usually recommended?

.. include:: /examples/generated/swiftui/Authenticate.snippet.flexible-sync-view-with-client-reset-handling.swift.rst

Take Some Action after the Client Reset Completes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

When the Sync error handler detects that a client reset must occur, the SDK 
executes the client reset using the behavior you specify in the 
``flexibleSyncConfiguration``'s ``clientResetMode`` parameter. However, the 
SwiftUI ``@AsyncOpen`` and ``@AutoOpen`` property wrappers do not detect 
that a client reset has occurred. An app that :ref:`opens a synced realm by 
switching on @AsyncOpen or @AutoOpen property wrapper states 
<ios-open-synced-realm-swiftui>` displays a progress spinner indefinitely. 
It never progresses past the ``.connecting`` state. The next time the user 
opens the app, it loads without issue.

> You can take some action in the ``afterClientReset`` block to handle this in

Reviewer: what do you mean by 'take some action'?

Author: I list out some possible options after this line. Unfortunately, I couldn't get a "good" solution for this. I wanted to find a way to gracefully handle it, but I tried a bunch of things and could only get sub-optimal solutions. So I left this intentionally vague. I'm hoping to get a recommendation from the SDK team to handle it gracefully, in which case I'll update this to a specific recommendation.

Reviewer: sounds good. let's leave this comment thread open in case there's some SDK team feedback and changes that you'd like reviewed.

Reviewer: or how about smthn like: 

```suggestion
You can perform additional logic in the ``afterClientReset`` block to handle this in
```

> your app. You might set a ``@State`` variable, modify its value in the 

Reviewer: ```suggestion
your app. You could set a ``@State`` variable, modify its value in the 
```

``afterClientResetBlock``, and use that to instruct the user to 

> re-open the app when a client reset occurs. Or you might force the 

Reviewer: ```suggestion
re-open the app when a client reset occurs. Or you could force the 
```

``@AsyncOpen`` or ``@AutoOpen`` View to reload using the updated realm.

Verify Client Reset Works as Expected
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> To test that your client reset implementation is working, you can trigger
> a client reset by terminating and re-enabling Device Sync.

Reviewer: ```suggestion
To test that your client reset implementation is working,
trigger a client reset by terminating and re-enabling Device Sync.
```

.. warning::

   Test client reset handling in a development app that does not have 

>    public users. Triggering a client reset using this method triggers it 

Reviewer: what do you mean by 'public users'? i'd consider alt phrase like 'real users', 'external users', etc.

>    public users. Triggering a client reset using this method triggers it 

Reviewer: ```suggestion
   public users. Causing a client reset using this method triggers it 
```

avoids repeating 'trigger'. feel free to adjust phrasing.

   for *all* app users. If your app does not yet have a functional client

>    reset handler, this can leave users in a bad app state.

Reviewer: confused by this sentence. 

> If your app does not yet have a functional client reset handler
how would the app not have a functional client reset handler if you're testing client reset handlers here? i imagine you're talking about different versions of the app or something similar. worth specifying here.

> this can leave users in a bad app state

can you expand on what you mean by a 'bad app state'?

Open your app on the device. Add some data to your app. Then, stop your 
app.

Follow the instructions in the Device Sync documentation to :ref:`terminating-realm-sync`.
You should see a ``TranslatorFatalError Error`` message in the App Services 
logs, indicating that Sync has terminated. Then, :ref:`re-enable-realm-sync`. 

When you run the app on your device again, it should trigger a client reset 
error. Verify that you see any relevant log messages that you added to your 
code while implementing the client reset logic, and that the device
resets the realm and successfully loads fresh data from the backend.

Repeat until you are satisfied with the client reset implementation in your

> SwiftUI app.

Reviewer: could you make this a numbered list?

For more information about the underlying error and recovery options, 
refer to: :ref:`ios-client-reset`.

Realm's SwiftUI property wrappers ``@AsyncOpen`` and ``@AutoOpen`` enable
you to handle a client reset error by adding a few things to your app:

- Add client reset handling information to the :swift-sdk:`flexibleSyncConfiguration() 
  <Extensions/User.html#/s:So7RLMUserC10RealmSwiftE25flexibleSyncConfigurationAC0B0V0F0VyF>`
- Force an ``@AsyncOpen`` or ``@AutoOpen`` view to reload after a client reset is complete

Add Client Reset Handling Information to the Flexible Sync Configuration
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The Swift SDK can handle a client reset with different behaviors depending
on the needs of your app. For a full list of client reset modes and behaviors,
refer to :ref:`ios-client-resets`.

To automatically handle a client reset, set the ``clientResetMode`` parameter 
in your :swift-sdk:`flexibleSyncConfiguration() 
<Extensions/User.html#/flexibleSyncConfiguration(clientResetMode:cancelAsyncOpenOnNonFatalErrors:)>`.
You can review the possible values and their behaviors in the 

> :swift-sdk:`ClientResetMode enum <Enums/ClientResetMode.html>`

Reviewer: ```suggestion
:swift-sdk:`ClientResetMode enum <Enums/ClientResetMode.html>`.
```

.. include:: /examples/generated/swiftui/Authenticate.snippet.flexible-sync-view-with-client-reset-handling.swift.rst

Take Some Action after the Client Reset Completes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

When the Sync error handler detects that a client reset must occur, the SDK 
executes the client reset using the behavior you specify in the 
``flexibleSyncConfiguration``'s ``clientResetMode`` parameter. However, the 
SwiftUI ``@AsyncOpen`` and ``@AutoOpen`` property wrappers do not detect 
that a client reset has occurred. An app that :ref:`opens a synced realm by 
switching on @AsyncOpen or @AutoOpen property wrapper states 
<ios-open-synced-realm-swiftui>` displays a progress spinner indefinitely. 
It never progresses past the ``.connecting`` state. The next time the user 
opens the app, it loads without issue.

You can take some action in the ``afterClientReset`` block to handle this in
your app. You could set a ``@State`` variable, modify its value in the 
``afterClientResetBlock``, and use that to instruct the user to 
re-open the app when a client reset occurs. Or you could force the 

> ``@AsyncOpen`` or ``@AutoOpen`` View to reload using the updated realm.

Reviewer: should 'View' be capitalized?
